kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandDecryptionManager.java

```diff
@@ -104,112 +154,117 @@ public CompletionStage<MemoryRecords> decrypt(@NonNull String topicName, int par
-            var decryptionVersion = decryptionVersion(topicName, partition, kafkaRecord);
-            if (decryptionVersion == EncryptionVersion.V1) {
-                ByteBuffer wrapper = kafkaRecord.value();
-                var edekLength = ByteUtils.readUnsignedVarint(wrapper);
-                ByteBuffer slice = wrapper.slice(wrapper.position(), edekLength);
-                var edek = serde.deserialize(slice);
-                edeks.add(edek);
-            }
-        }
-        return edeks;
-    }
+    /**
+     * Gets each record's encryption header (any any) and then resolves those edeks into
+     * {@link io.kroxylicious.filter.encryption.dek.Dek.Decryptor}s via the {@link #dekManager}
+     * @param topicName The topic name.
+     * @param partition The partition.
+     * @param records The records to decrypt.
+     * @return A stage that completes with a list of the DecryptState
+     * for each record in the given {@code records}, in the same order.
+     */
+    private CompletionStage<List<DecryptState<E>>> resolveAll(String topicName,
+                                                              int partition,
+                                                              MemoryRecords records) {
+        Serde<E> serde = dekManager.edekSerde();
+        var cacheKeys = new ArrayList<CacheKey<E>>();
+        var states = new ArrayList<DecryptState<E>>();
 
-    @NonNull
-    private BatchAwareMemoryRecordsBuilder decrypt(String topicName,
-                                                   int partition,
-                                                   @NonNull MemoryRecords records,
-                                                   @NonNull BatchAwareMemoryRecordsBuilder builder,
-                                                   @NonNull Map<E, AesGcmEncryptor> encryptorMap,
-                                                   @NonNull List<Integer> batchRecordCounts) {
-        int i = 0;
-        for (MutableRecordBatch batch : records.batches()) {
-            Integer batchRecordCount = batchRecordCounts.get(i++);
-            if (batchRecordCount == 0 || batch.isControlBatch()) {
-                builder.writeBatch(batch);
+        // Iterate the records collecting cache keys and decrypt states
+        // both cacheKeys and decryptStates use the list index as a way of identifying the corresponding
+        // record: The index in the list is the same as their index within the MemoryRecords
+        RecordStream.ofRecords(records).forEachRecord((batch, record, ignored) -> {
+            var decryptionVersion = decryptionVersion(topicName, partition, record);
+            if (decryptionVersion != null) {
+                ByteBuffer wrapper = record.value();
+                // TODO it's ugly passing a BiFunction like CacheKey::new
+                // should the wrapper just return a CacheKey directly
+                cacheKeys.add(decryptionVersion.wrapperVersion().readSpecAndEdek(wrapper, serde, CacheKey::new));
```

Comment:
I'm using a `BiFunction` in the `WrapperVersion.readSpecAndEdek()` and passing `CacheKey::new` because I don't want to make `WrapperVersion` depend directly on `CacheKey` (which feels like it's internal to how the cache works). I don't think this is ideal, but I couldn't see a cleaner way.

Reply:
Yep, I think bifunction is fine for now. `(Cipher id, EDEK)` is specific to the cache.